torchrl/envs/utils.py

```diff
@@ -181,34 +187,46 @@ def step_mdp(
             next_tensordict.update(out)
             return next_tensordict
         return out
-    out = tensordict.get("next").clone(False)
-    excluded = set()
-    if exclude_done:
-        excluded.add(done_key)
-    if exclude_reward:
-        excluded.add(reward_key)
-    if len(excluded):
-        out = out.exclude(*excluded, inplace=True)
-    td_keys = None
+
+    td = tensordict.exclude("next")
+    td_next = tensordict.get("next")
+
+    td_keys = td.keys(True, True)
+    td_next_keys = td_next.keys(True, True)
+
+    out = TensorDict({}, td_next.batch_size)
+
+    # Set the keys from next
+    excluded = {
+        done_key if exclude_done else None,
+        reward_key if exclude_reward else None,
+    }
+    for key in td_next_keys:
+        if key not in excluded:
+            _set_key(dest=out, source=td_next, key=key)
+
+    # Set the keys from root
+    if not exclude_action:
+        _set_key(dest=out, source=tensordict, key=action_key)
     if keep_other:
-        out_keys = set(out.keys())
-        td_keys = set(tensordict.keys()) - out_keys - excluded - {"next"}
-        if exclude_action:
-            td_keys = td_keys - {action_key}
-    elif not exclude_action:
-        td_keys = {action_key}
-
-    if td_keys:
-        # update does some checks that we can spare
-        # out.update(tensordict.select(*td_keys))
+        excluded = set.union(excluded, {action_key}, td_next_keys)
         for key in td_keys:
-            out._set(key, tensordict.get(key))
+            if key not in excluded:
+                _set_key(dest=out, source=tensordict, key=key)
+
     if next_tensordict is not None:
         return next_tensordict.update(out)
     else:
         return out
 
 
+def _set_key(dest, source, key):
+    dest._set(key, source.get(key))
+    if isinstance(key, tuple) and len(key) > 1:  # Setting the batch_sizes
+        for k in range(1, len(key)):
+            dest[key[:k]].batch_size = source[key[:k]].batch_size
```

Comment:
if two keys share the same root, we'll set the batch-time multiple times, which is time consuming

Reply:
yeah I know, i ll improve it